introduction_to_amazon_algorithms/jumpstart-foundation-models/text-generation-benchmarking/benchmarking/load_test.py

```diff
@@ -0,0 +1,344 @@
+                elif "generation" in data:
+                    data = data["generation"]
+                else:
+                    raise ValueError("Output data contains dictionary without recognized keys.")
+
+        text_inputs = self._text_inputs()
+        if data.startswith(text_inputs):
+            data = data[len(text_inputs) :]
+
+        return data
+
+    def num_words(self) -> int:
+        """The word count of the output sequence."""
+        return self._num_words(self.output_sequence())
+
+    def num_tokens(self, tokenizer: PreTrainedTokenizerBase) -> int:
+        """The token count of the output sequence."""
+        return len(tokenizer.encode(self.output_sequence()))
+
+    def _text_inputs(self) -> str:
+        if "inputs" in self.payload:
+            return self.payload["inputs"]
+        elif "text_inputs" in self.payload:
+            return self.payload["text_inputs"]
+        else:
+            raise ValueError("Expected input text keys are not detected in payload.")
+
+    @staticmethod
+    def _num_words(text: str) -> int:
+        return len(text.split())
+
+
+class BatchInvocationStatistics(NamedTuple):
+    """A NamedTuple holding start and stop times for a batch of endpoint predictions."""
+
+    time_utc_start: datetime.datetime
+    time_utc_end: datetime.datetime
+    num_invocations: int
+    results: List[PredictionResult]
+
+    def _duration_seconds(self) -> float:
+        """Computes the time in seconds of the batch load test."""
+        return (self.time_utc_end - self.time_utc_start).total_seconds()
+
+    def _throughput(self, values: List[int]) -> float:
+        """Computes the number of invocation responses per second."""
+        return sum(values) / self._duration_seconds()
+
+    def _throughput_robust(self, values: List[int]) -> float:
+        """Computes the median throughput of result values in units values per second to eliminate outliers."""
+        throughput_values = [
+            sum(values[: (i + 1)]) / (result.time_utc_end - self.time_utc_start).total_seconds()
+            for i, result in enumerate(self.results)
+        ]
+        return self._collect_statistics(throughput_values)["Maximum"]
+
+    def get_statistics(
+        self,
+        tokenizer: Optional[PreTrainedTokenizerBase] = None,
+        price_per_endpoint: Optional[float] = None,
+    ) -> Dict[str, Any]:
+        """Collect statistics on the number of input/output sequence words, the latency, and the latency per word."""
+        latency_per_word = self._collect_statistics(
+            [x.client_latency() / x.num_words() for x in self.results if x.num_words() > 0]
+        )
+        word_throughput_robust = self._throughput_robust([x.num_words() for x in self.results])
+        time_to_generate_1m_words = 1e6 / word_throughput_robust / 3600
+        statistics: Dict[str, Any] = {
+            "InputSequenceWords": self._collect_statistics([x.input_sequence_num_words() for x in self.results]),
```

Comment:
optional: consider making a dataclass to store these stats instead of a dictionary